src/cargo/core/compiler/build_context/target_info.rs

```diff
@@ -761,6 +760,12 @@ impl<'cfg> RustcTargetData<'cfg> {
                     .default_kind()
                     .into_iter()
                     .chain(p.manifest().forced_kind())
+                    .chain(
+                        p.manifest()
+                            .dependencies()
+                            .iter()
+                            .filter_map(|d| d.artifact()?.target()?.to_compile_kind()),
+                    )
```

Comment:
That's a great catch. Even though I don't know about the `force-kind` code as I don't think I created it, the example below helped to setup a test to reproduce it.

Strangely enough it seems to have surfaced another issue with artifacts not being handled if they are entirely in dependencies coming from a registry. Due to that I couldn't reproduce the panic shown above and from all I can tell, the setup I built here is correct.

Here is the progress so far: d318db1cb4b17a025601759cb4fc1193dd965849 .

Reply:
Here is what I could find out. The `pkg_set` does contain artifact information in its summaries after [download](https://github.com/rust-lang/cargo/blob/86dfd168ae58717fb8096466e455445ac2720157/src/cargo/ops/resolve.rs#L153). What's used for building unit dependencies though is the `resolve` information ends up with summaries that don't have artifact information [right after this call](https://github.com/rust-lang/cargo/blob/86dfd168ae58717fb8096466e455445ac2720157/src/cargo/ops/resolve.rs#L197).

The cause turned out to be a very real issue, as artifact information previously wasn't part of the registry which is the only source of information during the resolve stage. My solution is here: c9ab8f54dc24c34b81d5eacafa9f47c3b627ca02  .

An alternative could have been to try to backport dependency information from the `pkg_set` after downloading manifests, but that seemed like its a new mechanism which doesn't exist for a reason.

With the latest changes, I am able to reproduce the actual panic `no entry found for key', src/cargo/core/compiler/build_context/target_info.rs:821:40` and can work on fixing that.

Reply:
The fix provided here f8ee212ff03df6f0814a05b0978c98dfed300dde is no more than a basis for discussion as it's probably not correct.

The issue seems to be cycle between `target_data` and `resolve` information. The current approach somewhat ignores that and is happy with the succeeding test, but I have a feeling one could build examples that would fail during resolution (i.e. when `target_data` still lacks the targets from the complete dependency graph).

If the current way of doing this is indeed incorrect, I could imagine settling the cycle by redoing the resolution with the then change `target_data` until the solution is stable, while assuring it's not periodic and or otherwise never-ending.

I think the above shows that a proper implementation relies on advice by the reviewers, as I myself feel this shows signs of going off-track already.

Thanks for your support.